src/AzureExtension/DataManager/AzureDataManagerCache.cs

```diff
@@ -0,0 +1,296 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Dynamic;
+using DevHomeAzureExtension.Client;
+using DevHomeAzureExtension.DataManager;
+using DevHomeAzureExtension.DataModel;
+using DevHomeAzureExtension.DeveloperId;
+using Microsoft.TeamFoundation.Core.WebApi;
+using Microsoft.TeamFoundation.SourceControl.WebApi;
+using Microsoft.VisualStudio.Services.Account;
+using Microsoft.VisualStudio.Services.Account.Client;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.WebApi;
+using Serilog;
+
+namespace DevHomeAzureExtension;
+
+public partial class AzureDataManager
+{
+    private static readonly Uri VSAccountUri = new(@"https://app.vssps.visualstudio.com/");
+
+    private static readonly int PullRequestProjectLimit = 50;
+
+    private static readonly int PullRequestRepositoryLimit = 25;
+
+    public async Task UpdateDataForAccountsAsync(RequestOptions? options = null, Guid? requestor = null)
+    {
+        // A parameterless call will always update the data, effectively a 'force update'.
+        await UpdateDataForAccountsAsync(TimeSpan.MinValue, options, requestor);
+    }
+
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+    public async Task UpdateDataForAccountsAsync(TimeSpan olderThan, RequestOptions? options = null, Guid? requestor = null)
+#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
+    {
+        var cancellationToken = options?.CancellationToken.GetValueOrDefault() ?? default;
+        var requestorGuid = requestor ?? Guid.Empty;
+        var errors = 0;
+        var accountsUpdated = 0;
+        var accountsSkipped = 0;
+        var startTime = DateTime.Now;
+        Exception? firstException = null;
+
+        // Refresh option will clear all sync data, effectively forcing every row to be updated.
+        if (options is not null && options.Refresh)
+        {
+            using var tx = DataStore.Connection!.BeginTransaction();
+            try
+            {
+                _log.Debug("Clearing sync data.");
+                Organization.ClearAllSynced(DataStore);
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, "Failed clearing sync data.");
+                tx.Rollback();
+            }
+
+            tx.Commit();
+        }
+
+        var developerIds = DeveloperIdProvider.GetInstance().GetLoggedInDeveloperIdsInternal();
+        foreach (var developerId in developerIds)
```

Comment:
For a loop this long with this many tabs.  Please return early.  I know that the loop won't execute if developerIds is empty, but this is a big loop that spans many lihnes.

Reply:
I don't understand what I would be returning early for? If developerIds is empty loop doesn't execute and it proceeds to log that nothing was done, send appropriate events and exit. If I were to return early there, I'd have to replicate that code unnecessarily.

However, in looking at that line, it occurred to me that DeveloperIdProvider can throw  and if it were to throw here it would probably crash the extension, so I updated that code to have a try catch to prevent the crash and capture the error for reporting.

Reply:
> I don't understand what I would be returning early for? 
Less tabs.

well.  No.  Because even if you did check and return early, you would still need to use a foreach loop.

hmm.  Maybe separate some of the logic inside the loop to a method?  Basically, it is a lot of code for 3 tabs in.

Reply:
Was able to make a good change here and refactor out 40 lines of the innermost loop into self-contained code that updates all the organization information, this cleaned up that method a lot and the flow is much easier to read and the org update code is now also easier to read, so goodness there which I think addresses this issue.